src/main/scala/hydrozoa/node/server/Node.scala

```diff
@@ -207,78 +216,187 @@ class Node(
-        log.info(s"Settlement tx: $serializedTx")
-        os.write.append(txDump, "\n" + serializedTx)
-
-        val Right(settlementTxId) = cardano.submit(settlementTx)
-
-        log.info(s"Settlement tx submitted: $settlementTxId")
-
-        multisigEventManager.map(_.handleSettlementTx(settlementTx, settlementTxId))
-
-        Right(serializedTx)
-
-    def produceFinalBlock: Either[String, String] =
-
-        // Block
-        val nextMajorVersion: Int = headStateManager.currentMajorVersion + 1
-        val block: Block = finalDummyBlock(nextMajorVersion)
-
-        // Tx
-        val recipe = FinalizationRecipe(nextMajorVersion)
-
-        val Right(finalizationTxDraft: FinalizationTx) =
-            finalizationTxBuilder.buildFinalizationTxDraft(recipe)
-
-        // Consensus
-        val ownWit: TxKeyWitness = signTx(finalizationTxDraft.toTx, ownKeys._1)
-        val ackFinalCombined: Set[AckFinalCombined] = network.reqFinal(block)
-        // TODO: broadcast ownWit
-
-        // Sign and submit
-        val wits: Set[TxKeyWitness] = ackFinalCombined.map(_.finalization) + ownWit
-        val finalizationTx: L1Tx = wits.foldLeft(finalizationTxDraft.toTx)(addWitness)
-        val serializedTx = serializeTxHex(finalizationTx)
-        log.info(s"Finalization tx: $serializedTx")
-        os.write.append(txDump, "\n" + serializedTx)
-
-        val Right(finalizationTxId) = cardano.submit(finalizationTx)
-        log.info(s"Finalization tx submitted: $finalizationTxId")
-
-        // TODO: temporary: handle the event
-        multisigEventManager.map(_.handleFinalizationTx(finalizationTx, finalizationTxId))
-
-        Right(serializedTx)
+    /** Manually triggers next block creation procedure.
+      * @param nextBlockFinal
+      * @return
+      */
+    def handleNextBlock(nextBlockFinal: Boolean): Either[String, String] =
+        
+        // 1. Initialize the variables and arguments.
+        // (a) Let block be a mutable variable initialized to an empty BlockL2
+        // instead on block we use mutable parts and finalize the block
+        // at the end using the block builder
+        val txValid, wdValid: mutable.Set[TxId] = mutable.Set.empty
+        val eventsInvalid: mutable.Set[(TxId, MempoolEventTypeL2)] = mutable.Set.empty
+        var depositsAbsorbed: Set[OutputRef[L1]] = Set.empty
+
+        // (b) Let previousBlock be the latest block in blocksConfirmedL2
+        val prevHeader = state.asOpen(_.l2Tip.blockHeader)
+
+        // (c) Let previousMajorBlock be the latest major block in blocksConfirmedL2
+        //val previousMajorBlock = state.asOpen(_.l2LastMajor)
+
+        // (d) Let utxosActive be a mutable variable initialized to stateL2.utxosActive
+        // for now (and probably this is legit) we use utxosActive within L2 ledger
+        // var utxosActive: UTxOs = state.asOpen(_.utxosActive)
+        // TODO: do we need to clone the ledger for block creation?
+        val stateL2 = state.asOpen(_.stateL2)
```

Comment:
The head state's confirmed L2 active utxo set should _never_ be modified when you create a new block.
The L2 effect of a new block only happens when it is confirmed -- all peer signatures have been collected.

Unless you have another way to achieve this, I think you do need to clone the L2 active utxo set.

Reply:
Theoretically, we could get fancy with some "copy on write" virtualization that reuses the confirmed L2 active utxo set (immutably) and overlays the added/deleted L2 utxos on top of it...

Reply:
On the other hand, we are already keeping track of the `utxosAdded` and `utxosWithdrawn`, so perhaps the overlay is not so far fetched...

Reply:
For extra efficiency, `utxosRemovedTxs` should only track the utxos spent from `utxosActiveConfirmed`. When utxos from `utxosAddedTxs` are spent, we can just directly remove them from that set.

Reply:
For efficiency, when L2 transactions (but not withdrawals!) spend utxos from `utxosAddedTxs`, we can just directly remove those utxos from that set.